ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java

```diff
@@ -478,58 +493,101 @@ private static ComparisonTarget extractWhereClauseTarget(final ComparisonExpress
     }
   }
 
-  private static boolean isSelectStar(final List<SelectItem> selects) {
+  private static boolean isSelectStar(final Select select) {
+    final List<SelectItem> selects = select.getSelectItems();
     return selects.size() == 1 && selects.get(0) instanceof AllColumns;
   }
 
-  private static Result handleSelects(
+  private static List<List<?>> handleSelects(
       final Result input,
       final ConfiguredStatement<Query> statement,
       final KsqlExecutionContext executionContext,
-      final Analysis analysis
+      final Analysis analysis,
+      final Builder schemaBuilder
   ) {
-    final List<SelectItem> selectItems = statement.getStatement().getSelect().getSelectItems();
-    if (input.rows.isEmpty() || isSelectStar(selectItems)) {
-      return input;
-    }
-
-    final LogicalSchema.Builder schemaBuilder = LogicalSchema.builder();
-    schemaBuilder.keyColumns(input.schema.key());
+    final LogicalSchema outputSchema = schemaBuilder.build();
 
-    final ExpressionTypeManager expressionTypeManager = new ExpressionTypeManager(
-        input.schema,
-        executionContext.getMetaStore()
-    );
-
-    for (int idx = 0; idx < analysis.getSelectExpressions().size(); idx++) {
-      final SelectExpression select = analysis.getSelectExpressions().get(idx);
-      final SqlType type = expressionTypeManager.getExpressionSqlType(select.getExpression());
-      schemaBuilder.valueColumn(select.getName(), type);
+    final LogicalSchema intermediateSchema;
+    final BiFunction<Struct, GenericRow, GenericRow> preSelectTransform;
+    if (outputSchema.key().isEmpty()) {
+      intermediateSchema = input.schema;
+      preSelectTransform = (key, value) -> value;
+    } else {
+      // SelectValueMapper requires the key fields in the value schema :(
+      schemaBuilder.valueColumns(outputSchema.key());
```

Comment:
This doesn't seem right to me. `schemaBuilder` is the schema of the output. IIUC what we're trying to do here is add the key columns to the input so we can use SelectValueMapper to project them into the output (since SelectValueMapper only knows about the value and not the key). So shouldn't we be adding the key fields to the schema of the input.

For example, what if my table has schema:
`X INT, Y INT, Z INT`

And I project:
`X, ROWKEY`

Won't this cause me to mess up the final projection? Since intermediate schema would be;
`X, ROWKEY` when it should be `X, Y, Z, ROWKEY`

Do we have a test like this?

Reply:
You're right! Good catch. Thanks man. Fixed.